draft-ietf-quic-http.md

```diff
@@ -1238,9 +1232,20 @@ MAX_PUSH_ID frame ({{frame-max-push-id}}). A client MUST treat receipt of a
 PUSH_PROMISE frame that contains a larger Push ID than the client has advertised
 as a connection error of H3_ID_ERROR.
 
-A server MUST NOT use the same Push ID in multiple PUSH_PROMISE frames. A client
-MUST treat receipt of a Push ID which has already been promised as a connection
-error of type H3_ID_ERROR.
+A server MAY use the same Push ID in multiple PUSH_PROMISE frames. If so, the
+decompressed request header sets MUST contain the same fields in the same
+order, and both the name and and value in each field MUST be exact
+matches. If a client receives a Push ID that has already been promised
+and detects a mismatch, it MUST respond with a connection error of type
+H3_GENERAL_PROTOCOL_ERROR. If the decompressed header sets match exactly, the
+client MUST ignore the duplicate PUSH_PROMISE frame.
```

Comment:
```suggestion
client SHOULD associate the pushed content with each stream on which a PUSH_PROMISE was received.
```
I don't think "MUST ignore" is what we want.  It's not an error case, it's a client optimization.

Reply:
Hm. I agree that it's an error, but I'm not sure what it means to associate pushed content with the request stream that received the PUSH_PROMISE. There is only going to be a single push stream with the specified push id, of course. And as such, only a single request can be directly associated with that push stream. Of course, because of caching, other requests can be satisfied by the cached response. But that's at a layer above the streams, right?

(in chrome, I don't think we have any mapping between pushed streams and the request stream that promised the pushes so maybe we're "doing it wrong"?)

Reply:
HTTP/2 associates each push with the (single) request where the server promised it.  HTTP/3 allows a push to be associated with multiple requests.

> There is only going to be a single push stream with the specified push id, of course. **And as such, only a single request can be directly associated with that push stream.**

This is the difference -- one need not follow from the other.  HTTP/3 allows a response to be promised on multiple streams, implying that you'll potentially want it as a dependency for each of those things.  That means it's intended to be possible for that push to satisfy several client requests (from different tabs, or threads, or iframes, or whatever).

Now, if the response is cacheable, you're right -- the HTTP cache will probably satisfy those other requests and it doesn't have to reach the push cache.  But the pushed response need not be cacheable; it can nonetheless be used multiple times if promised multiple times.

In WinInet, we had a concept of a navigation that was passed down from the browser, and the push cache was at the navigation level rather than the connection level.  It was part of whichever navigation included the stream that received the promise.  Hypothetically, if something were promised on several streams from different navigations, HTTP/3 would enable putting that response in both navigations' push caches.

Since Chrome can split navigations across connections, doesn't having the push cache at the connection level means you might miss that something was already pushed because it was promised on a different connection than Chrome was going to use for the request?

Reply:
Hm. I think there is a difference in philosophy here. 

> HTTP/2 associates each push with the (single) request where the server promised it. HTTP/3 allows a push to be associated with multiple requests.

Totally agree with this! If a client makes requests A and B, the server can promise to push resource C on both A and B. This is clearly true.

But at the HTTP/3 layer (not at the web browsing layer) there is no notion of navigation or of subsequent requests being "associated" with earlier HTTP/3 streams (on which a resource may have been promised). Or at least I don't think there is.

My understanding of the motivation for allowing multiple pushes in HTTP/3 is not because we want to allow multiple subsequent requests for the same resources to be associated with the same pushed stream. Instead it was because HTTP/2 is in-order and HTTP/3 is not. In HTTP/2 if a client requested A and B and the server knew that both referenced resource C, it could push C a single time and be confident that requests initiated by the reference to C in the bodies of A and B would both come *after* the promise of C. However in HTTP/3 this is not the case. Since the promise for C is delivered on the request stream there is no ordering guarantee that the push (on A) will arrive at the client before the body of B (which contains the reference to C) arrives. As such, if the server only wants to push C once, it needs to make sure both promises are know.

But this, to me, is orthogonal to the issue of associating multiple subsequent requests with the same push data down at the HTTP/3 layer because each of these requests are somehow associated with the original requests A and B.

Am I thinking about this the wrong way?

> Since Chrome can split navigations across connections, doesn't having the push cache at the connection level means you might miss that something was already pushed because it was promised on a different connection than Chrome was going to use for the request?

While Chrome does store pushes in a per-connection cache (which is really just the pushed streams themselves) it also has a "global" index of "promises by URL" which allows a request which would, by default, go on one connection to be satisfied by a promise from another connection (which might otherwise be going away, for example).

Reply:
Yes, I think we've found a conceptual difference.  You're looking at the promise purely as a way to forestall requests for that resource, and the fact that it happens to be on a given stream is only because the server thinks there might be something on that stream where you need to know about the push first.  But handling of push in Chrome is global -- if you request A and B, I promise X on A, and B references X, you'll use the pushed object if you know about it.  I agree, in this implementation style, the first PUSH_PROMISE frame to arrive carries all the information you need, and any subsequent ones are basically irrelevant.

But RFC7540 says that "Pushed responses are always associated with an explicit request from the client."  How (or whether, obviously!) that association is used is implementation-dependent, but I think for consistency's sake we need to provide that association in the document.  (As far as I know, the only thing H2 explicitly used this association for internally was for prioritization -- pushed streams depend on the associated stream.)

Do we need a separate editorial issue to expand on this a bit in the document, or is it there and we're just reading through different filters?

Reply:
> Yes, I think we've found a conceptual difference. You're looking at the promise purely as a way to forestall requests for that resource, and the fact that it happens to be on a given stream is only because the server thinks there might be something on that stream where you need to know about the push first. But handling of push in Chrome is global -- if you request A and B, I promise X on A, and B references X, you'll use the pushed object if you know about it. I agree, in this implementation style, the first PUSH_PROMISE frame to arrive carries all the information you need, and any subsequent ones are basically irrelevant.
> 
> But RFC7540 says that "Pushed responses are always associated with an explicit request from the client." How (or whether, obviously!) that association is used is implementation-dependent, but I think for consistency's sake we need to provide that association in the document. (As far as I know, the only thing H2 explicitly used this association for internally was for prioritization -- pushed streams depend on the associated stream.)

Interesting! I read the language in RFC7540 differently, I think. It says "Pushed responses are always associated with an explicit request from the client.  The PUSH_PROMISE frames sent by the server are sent on that explicit request's stream. " To me, this language exists to prevent unilateral pushes by the server. In other words, if there are no client-initiated requests outstanding, the server may not push any resources. 

But if I understand you correctly, you read it as enabling implementations to make use of this association for more than simply preventing unilateral pushes. (Presumably this association could be used by an implementation to "hide" the pushed resource from a subsequent request not associated with the original request, but to "show" it to a subsequent request that *was* associated with the original request? I would not have thought this was "a thing".) Fascinating.

> Do we need a separate editorial issue to expand on this a bit in the document, or is it there and we're just reading through different filters?

This is a good question :) I think some editorial exposition on this front would be welcome! I would be very curious to hear what other folks think of this issue.

In any case, backing up to the change you proposed:

> client SHOULD associate the pushed content with each stream on which a PUSH_PROMISE was received.

I think I'm comfortable with this text even though I'm slightly uneasy about what "associate" means here. But that's pretty orthogonal to this PR since it existed in the DUPLICATE_PUSH language as well, so I'll go ahead and apply your suggestion.

Thanks for the great discussion!